source/material_model/damage_rheology.cc

```diff
@@ -0,0 +1,2027 @@
+          else if (grain_size_change / grain_size > 0.1 || grain_size_growth / grain_size > 0.5
+                   || grain_size_reduction / grain_size > 0.5)
+            {
+              grain_size_change = 0.0;
+              time -= grain_growth_timestep;
+              grain_growth_timestep /= 2.0;
+            }
+
+          grain_size += grain_size_change;
+          current_composition[field_index] = grain_size;
+
+          if (grain_size < 0)
+            {
+              std::cout << "Grain size smaller 0:  " << grain_size << " ," << grain_size_growth
+                        << " ," << grain_size_reduction << ", timestep: " << grain_growth_timestep << "! \n ";
+              break;
+            }
+        }
+      while (time < timestep);
+
+      // reduce grain size to recrystallized_grain_size when crossing phase transitions
+      // if the distance in radial direction a grain moved compared to the last time step
+      // is crossing a phase transition, reduce grain size
+
+      // TODO: recrystallize first, and then do grain size growth/reduction for grains that crossed the transition
+      // in dependence of the distance they have moved
+      double phase_grain_size_reduction = 0.0;
+      if (this->introspection().name_for_compositional_index(field_index) == "olivine_grain_size"
+          &&
+          this->get_timestep_number() > 0)
+        {
+          // check if material has crossed any phase transition, if yes, reset grain size
+          if (crossed_transition != -1)
+            if (recrystallized_grain_size[crossed_transition] > 0.0)
+              phase_grain_size_reduction = grain_size - recrystallized_grain_size[crossed_transition];
+        }
+      else if (this->introspection().name_for_compositional_index(field_index) == "pyroxene_grain_size")
+        {
+          phase_grain_size_reduction = 0.0;
+        }
+
+      if (grain_size < 5.e-6)
+        {
+          std::cout << "Grain size is " << grain_size << "! It needs to be larger than 5e-6.\n";
+          grain_size = 5e-6;
+        }
+
+      return grain_size - original_grain_size - phase_grain_size_reduction;
+    }
+
+
+
+    template <int dim>
+    double
+    DamageRheology<dim>::
+    diffusion_viscosity (const double                  temperature,
+                         const double                  pressure,
+                         const std::vector<double>    &composition,
+                         const SymmetricTensor<2,dim> &strain_rate,
+                         const Point<dim>             &position) const
+    {
+      const SymmetricTensor<2,dim> shear_strain_rate = strain_rate - 1./dim * trace(strain_rate) * unit_symmetric_tensor<dim>();
+      const double second_strain_rate_invariant = std::sqrt(std::abs(second_invariant(shear_strain_rate)));
+
+      // TODO: make this more general, for more phases we have to average grain size somehow
+      // TODO: default when field is not given & warning
```

Comment:
We should decide what we want to do if there is no grain size field. There are a number of places that ask for the name of the field explicitly, and if it's not there, I would just crash with a Assert.